lib/pg.rb

```diff
@@ -5,141 +5,161 @@
-	if RUBY_ENGINE=="truffleruby" && !defined?(Truffle::WarningOperations)
-		module TruffleFixWarn
-			def warn(str, category=nil)
-				super(str)
-			end
-		end
-		Warning.extend(TruffleFixWarn)
-	end
-
-	# Ruby-3.4+ prints a warning, if bigdecimal is required but not in the Gemfile.
-	# But it's a false positive, since we enable bigdecimal depending features only if it's available.
-	# And most people don't need these features.
-	def self.require_bigdecimal_without_warning
-		oldverb, $VERBOSE = $VERBOSE, nil
-		require "bigdecimal"
-	ensure
-		$VERBOSE = oldverb
-	end
+  # Is this file part of a fat binary gem with bundled libpq?
+  # This path must be enabled by add_dll_directory on Windows.
+  gplat = Gem::Platform.local
+  bundled_libpq_path = Dir[File.expand_path("../ports/#{gplat.cpu}-#{gplat.os}*/lib", __dir__)].first
+  if bundled_libpq_path
+    POSTGRESQL_LIB_PATH = bundled_libpq_path
+  else
+    # Try to load libpq path as found by extconf.rb
+    begin
+      require "pg/postgresql_lib_path"
+    rescue LoadError
+      # rake-compiler doesn't use regular "make install", but uses it's own install tasks.
+      # It therefore doesn't copy pg/postgresql_lib_path.rb in case of "rake compile".
+      POSTGRESQL_LIB_PATH = false
+    end
+  end
+  POSTGRESQL_LIB_PATH.freeze
+
+  add_dll_path = proc do |path, &block|
+    if RUBY_PLATFORM =~/(mswin|mingw)/i && path
+      BUNDLED_LIBPQ_WITH_UNIXSOCKET = false
+      begin
+        require 'ruby_installer/runtime'
+        RubyInstaller::Runtime.add_dll_directory(path, &block)
+      rescue LoadError
+        old_path = ENV['PATH']
+        ENV['PATH'] = "#{path};#{old_path}"
+        block.call
+        ENV['PATH'] = old_path
+      end
+    else
+      # libpq is found by a relative rpath in the cross compiled extension dll
+      # or by the system library loader
+      block.call
+      BUNDLED_LIBPQ_WITH_UNIXSOCKET = RUBY_PLATFORM=~/linux/i && PG::IS_BINARY_GEM
+    end
+  end
+
+  # Add a load path to the one retrieved from pg_config
+  add_dll_path.call(POSTGRESQL_LIB_PATH) do
+    begin
+      # Try the <major>.<minor> subdirectory for fat binary gems
+      major_minor = RUBY_VERSION[ /^(\d+\.\d+)/ ] or
+        raise "Oops, can't extract the major/minor version from #{RUBY_VERSION.dump}"
+      require "#{major_minor}/pg_ext"
+    rescue LoadError => error1
+      begin
+        require 'pg_ext'
+      rescue LoadError => error2
+        msg = <<~EOT
+          pg's C extension failed to load:
+            #{error1}
+            #{error2}
+        EOT
+        if msg =~ /GLIBC/
+          msg += <<~EOT
+
+            The GLIBC version of this system seems too old. Please use the source version of pg:
+                gem uninstall pg --all
+                gem install pg --platform ruby
+            or in your Gemfile:
+                gem "pg", force_ruby_platform: true
+            See also: https://github.com/ged/ruby-pg/blob/master/README.md#source-gem
```

Comment:
In URLs, the `-` branch resolves to the HEAD branch, ensuring that this link will continue to work even if the HEAD branch changes its name.

```suggestion
            See also: https://github.com/ged/ruby-pg/blob/-/README.md#source-gem
```

Reply:
Sounds good! Alternatively maybe a link to https://deveiate.org/code/pg/README_md.html#label-Source+gem ? Don't have a strong opinion.